doc/read-the-docs-site/adr/0006-indexer-resuming-strategy.rst

```diff
@@ -45,16 +45,88 @@ Decision
   In the current Marconi interface, it is actually just ``securityParam + 1`` as we assume that all rollbackable blocks *can* be fully stored on disk because of the disk flush.
   We assume that all the data contained in the indexer is consistent for all the point returned by an indexer.
 
-* We will have each indexer start the node-to-client chain-sync protocol at different points in time.
-  However, the ``Coordinator`` will make sure that all indexers are synced until a common point, and advance at the same speed (i.e. they can only request the next block once all the indexers have finished processing the current block).
+* We will start a single node chain-sync client for all indexers instead of one node chain sync client per indexer.
 
-* We will use the full set of points provided by ``resumableFromStorage`` of an indexer as resuming points for the node-to-client chain-sync protocol.
-  These points need to ordered in descending order so that protocol can priotise the selection of the latest ones.
+* We will implement the following resuming logic.
+
+  In order to start the chain-sync client, we need to provide to provide intersection points (or resuming points).
+  Intersection points represent the points from which we want to start syncing from the local node.
+  We will provide as intersection points the points returned from the following pseudo-code:
+
+  .. code-block:: haskell
+
+    -- | Query the resumable points of each indexer then select the resumable
+    -- points of the indexer that has the lowest immutable point. If two indexer
+    -- have the same lowest immutable point, select the one which has the lowest
+    -- overall point (immutable or not).
+    selectIntersectionPoints indexers = do
+      resumingPoints <- mapM queryResumingPoints indexers
+      head $ sortOn (\rps -> (maximum $ filter isImmutablePoint rps, maximum rps)) resumingPoints
+
+  Next, we need a function which processes each ``ChainSyncEvent`` returned by the chain-sync client for each indexer.
+  The pseudo-code is as follows:
+
+  .. code-block:: haskell
+
+    applyChainSyncEvent event indexer = do
+      case event of
+        lastSyncPoints <- getLastSyncPoints indexer
+        Resume point -> do
+          maybePoint <- getPointFromSlotNo (slotNo point) indexer
+          case maybePoint of
+            Nothing ->
+              pure () -- Do nothing
+            Just _ ->
+              rewind point indexer
+        RollBackward point -> do
+          rewind point indexer
+        RollForward block point -> do
+          lastSyncPoints <- getLastSyncPoints indexer
```

Comment:
we may want to avoid recomputing it after each new event.

Reply:
Good point. I'll change that.

Reply:
Mmm.. But at each event, we still need to update the database right? So querying won't add too much overhead, afaik? The main issue I see is that it complexifies the implementation because we need to keep track of the last sync points by only keeping K points, and removing old points after each update.

Reply:
No, we don't. The list of points we consider for resuming is _almost_ immutable: as long as we drain, the database shouldn't change.
I say _almost_ because we may want to remove points from this list when we catch up, to ease the processing.

Reply:
I'll revisit.